### DISCUSSION TITLE
How to update child state from parent reducer by trigger child action

### DISCUSSION THREAD

**Discussion by vvisionnn** [pointfreeco/swift-composable-architecture#1597] [Q&A] [answered]
Hey guys, I have been using composable architecture with UIKit for few months, and I still cannot find a better way to figure it out that how to update the child state from the parent reducer

things is this, think we have a `Parent` Feature and a `Child` Feature, and for the `Parent` and `Child`, both of them got an action called **`refresh`** like below

```swift
struct Parent: ReducerProtocol {
    struct State: Equatable {
        ......
        var child: Child.State?
    }

    enum Action: Equatable {
        case refresh
    }

......
    var Body: some ReducerProtocol<State, Action> { Reduce { state, action in
        switch action {
        case .refresh:
            /* some parent state update */
            // then update the child state by trigger the action called `Child.Action.refresh`
        return .none
    }
}

struct Child: ReducerProtocol {
    @Dependency(\.database) var db

    struct State: Equatable {
        ......
    }

    enum Action: Equatable {
        case refresh
    }

......
    var Body: some ReducerProtocol<State, Action> { Reduce { state, action in
        switch action {
        case .refresh:
            /* some child state update from db */
            state.a = db.findA()
            state.b = db.findB()
        return .none
    }
}
```

things is, some times there may have no directly connection from parent state to child state, which means that I cannot using the compute property to calculate the new child state

like we have modified some database rows out of the `Parent`, but we want to update the whole parent state because **some** (not one) child states inside parent should reflect the database value which we have modified before, the child refresh action will handle this but I cannot trigger it from parent refresh action

I have discovered several ways to solve this,
1. first is I will modify the child state directly from the parent refresh action, that will work but I cannot guarantee that will cover all the  property inside child state
2. second is I will trigger the action by create a new reducer then call the `reduce(into:action:)` method, like `Child().reduce(&state.child, action: .refresh)` that works, in some way, but if the child refresh returns the side effect, like `.merge([.refreshA, .refreshB])` or just `.refreshC`, I can only loop refresh it, or somedays before that I just missed it

I think non of these solved my problem in some way, and I just want to know if there is an elegant and simple method to help me fix this, if so that would be helpful

**Comment by Alex293** [comment, marked as answer]
To make the parent send actions to the child, the parent feature should be composed with the child feature. 

For example you can add a case child(ChildFeature.Action) to the parent actions. Make the parent reduce func or body run the child with ifLet or Scope. Then In your parent refresh action return Effect(value: .child(.refresh)) instead of .none. 

There are multiples ways of doing this depending on the relation of you features. If they are unrelated you could have a third reducer that compose your features and relay the action

**Reply by vvisionnn** [reply]
Thanks! that is correct,I don't know why that I have tried this before and won't work, maybe something I have misconfigured and I don't know, this is really helpful, thanks @Alex293 !
